drivers/iommu/riscv/iommu.c

```diff
@@ -831,612 +1096,692 @@ static struct iommu_group *riscv_iommu_device_group(struct device *dev)
-	} while (1);
 
-	return IRQ_HANDLED;
-}
+/************************\
+* Translation mode setup *
+\************************/
 
-/* Wait for DDTP.BUSY to be cleared and return latest value */
 static u64 riscv_iommu_get_ddtp(struct riscv_iommu_device *iommu)
 {
 	u64 ddtp;
 	cycles_t end_cycles = RISCV_IOMMU_TIMEOUT + get_cycles();
 
+	/* Wait for DDTP.BUSY to be cleared and return latest value */
 	do {
-		ddtp = riscv_iommu_readq(iommu, RIO_REG_DDTP);
-		if (!(ddtp & RIO_DDTP_BUSY))
+		ddtp = riscv_iommu_readq(iommu, RISCV_IOMMU_REG_DDTP);
+		if (!(ddtp & RISCV_IOMMU_DDTP_BUSY))
 			break;
 		cpu_relax();
 	} while (get_cycles() < end_cycles);
 
 	return ddtp;
 }
 
-/* Enable IOMMU Translation Mode. */
-static int riscv_iommu_enable(struct riscv_iommu_device *iommu, unsigned mode)
+static void riscv_iommu_ddt_cleanup(struct riscv_iommu_device *iommu)
 {
-	u64 ddtp;
+	if (iommu->ddtp) {
+		// TODO: teardown whole tree.
+		if (iommu->ddtp_in_iomem) {
+			iounmap((void*) iommu->ddtp);
+		} else
+			free_page(iommu->ddtp);
+		iommu->ddtp = 0;
+	}
+}
+
+static int riscv_iommu_enable(struct riscv_iommu_device *iommu, unsigned requested_mode)
+{
+	struct device *dev = iommu->dev;
+	u64 ddtp = 0;
+	u64 ddtp_paddr = 0;
+	unsigned mode = requested_mode;
+	unsigned mode_readback = 0;
 
 	ddtp = riscv_iommu_get_ddtp(iommu);
-	if (ddtp & RIO_DDTP_BUSY)
+	if (ddtp & RISCV_IOMMU_DDTP_BUSY)
 		return -EBUSY;
 
-	/* Disallow state transtion from xLV to xLV. */
-	switch (FIELD_GET(RIO_DDTP_MODE, ddtp)) {
-	case RIO_DDTP_MODE_BARE:
-	case RIO_DDTP_MODE_OFF:
+	/* Disallow state transtion from xLVL to xLVL. */
+	switch (FIELD_GET(RISCV_IOMMU_DDTP_MODE, ddtp)) {
+	case RISCV_IOMMU_DDTP_MODE_BARE:
+	case RISCV_IOMMU_DDTP_MODE_OFF:
 		break;
 	default:
-		if ((mode != RIO_DDTP_MODE_BARE) && (mode != RIO_DDTP_MODE_OFF))
+		if ((mode != RISCV_IOMMU_DDTP_MODE_BARE) && (mode != RISCV_IOMMU_DDTP_MODE_OFF))
 			return -EINVAL;
 		break;
 	}
 
-	/* Verify supported DDTP.MODE values */
+ retry:
 	switch (mode) {
-	case RIO_DDTP_MODE_BARE:
-	case RIO_DDTP_MODE_OFF:
-		if (iommu->ddtp) {
-			// TODO: teardown whole tree.
-			free_pages(iommu->ddtp, 0);
-			iommu->ddtp = 0;
-		}
-		ddtp = FIELD_PREP(RIO_DDTP_MODE, mode);
+	case RISCV_IOMMU_DDTP_MODE_BARE:
+	case RISCV_IOMMU_DDTP_MODE_OFF:
+		riscv_iommu_ddt_cleanup(iommu);
+		ddtp = FIELD_PREP(RISCV_IOMMU_DDTP_MODE, mode);
 		break;
-	case RIO_DDTP_MODE_1LVL:
-	case RIO_DDTP_MODE_2LVL:
-	case RIO_DDTP_MODE_3LVL:
-		if (!iommu->ddtp)
-			iommu->ddtp = get_zeroed_page(GFP_KERNEL);
+	case RISCV_IOMMU_DDTP_MODE_1LVL:
+	case RISCV_IOMMU_DDTP_MODE_2LVL:
+	case RISCV_IOMMU_DDTP_MODE_3LVL:
+		if (!iommu->ddtp) {
+			/*
+			 * We haven't initialized ddtp yet, since it's WARL make
+			 * sure that we don't have a hardwired PPN field there
+			 * that points to i/o memory instead.
+			 */
+			riscv_iommu_writeq(iommu, RISCV_IOMMU_REG_DDTP, 0);
+			ddtp = riscv_iommu_readq(iommu, RISCV_IOMMU_REG_DDTP);
```

Comment:
Using riscv_iommu_get_ddtp() will also cover handling BUSY flag after DDTP write.

Reply:
I wanted to avoid the loop (and we are going to set the mode later on anyway) but ok, I'll switch it to riscv_iommu_get_ddtp(iommu);